zerver/tests/test_subs.py

```diff
@@ -2275,8 +2289,11 @@ def test_bulk_subscribe_many(self) -> None:
                     streams,
                     dict(principals=ujson.dumps([self.test_email])),
                 )
-        # Make sure we don't make O(streams) queries
-        self.assert_length(queries, 21)
+        # For each stream, we have to check for subscribed guest users,
+        # so the number of queries performed have a complexity of O(stream),
+        # more specifically proportional to two times of streams because we
+        # perform two queries  for each stream in `public_stream_user_ids`
+        self.assert_length(queries, 60)
```

Comment:
I hope this comment conveys what's happening here. I can't think of anyway by which we can reduce this count because we have to deal with each stream for subscribed users individually. Also, I can't see any real situation where we deal with this many numbers of the stream in our web app.

Reply:
This is a real issue that we'll need to deal with.  It is possible to manage; we just need to do a bulk query (which may require some significant refactoring, unfortunately).  I'm not going to merge this commit; let's open it as a separate issue since I think we'll need some more extended thinking and discussion, and we don't strictly have to fix this for the feature to be shippable.